Kafka-5160; KIP-98 Broker side support for TxnOffsetCommitRequest

apurvam opened the pull request:
This patch adds support for the `TxnOffsetCommitRequest` added in KIP-98. Desired handling for this request is [described here](https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#bookmark=id.55yzhvkppi6m) . 

The functionality includes handling the stable state of receiving `TxnOffsetCommitRequests` and materializing results only when the commit marker for the transaction is received. It also handles partition emigration and immigration and rebuilds the required data structures on these events.

Tests are included for all the functionality.

Commit:
WIP

Commit:
WIP commit:

 1. Able to write txn offset commits to the log and delay
materialization until the commit marker appears.
 2. Able to recover these commits on partition load.

Todo:

 1. Tests for the above.
 2. Materialize or drop cached offset commits when the transaction
marker is received.

Commit:
Small simplification

Commit:
fix indentation

Commit:
Code complete barring integration. Now to add unit tests

Commit:
Added the first test cases

Commit:
Completed the functional test cases

apurvam commented:
cc @hachikuji @guozhangwang

Commit:
WIP

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/RecordBatch.java:
```diff
@@ -55,6 +55,12 @@
+     * For offset commit messages, we need to skip the sequence checks since sequence numbers are not
+     * generated on the producer for these messages.
+     */
+    int SKIP_SEQUENCE_CHECK = -3;
```
A bit unfortunate that we need this. If we want to support this, we'll have to add something to `LogValidator` to ensure that users aren't using this as well. One alternative that comes to mind is to add an option to `Log` to disable the sequence check if it's an internal topic (or maybe just if it's __consumer_offsets). Then we could use `NO_SEQUENCE`. Seems a bit nicer to keep this bit of ugliness in the code and not the protocol.

apurvam replied:
I agree. I changed the code to have the `Log` exempt writes to the offsets topic from sequence number validation.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -410,32 +413,41 @@ class GroupCoordinator(val brokerId: Int,
       if (group.is(Dead)) {
         responseCallback(offsetMetadata.mapValues(_ => Errors.UNKNOWN_MEMBER_ID))
-      } else if (generationId < 0 && group.is(Empty)) {
+      } else if ((generationId < 0 && group.is(Empty)) || (producerId != RecordBatch.NO_PRODUCER_ID)) {
```
One question I'm wondering is whether we should allow groups to simultaneously support normal consumer offset commits and transactional offset commits. I can't think of a valid use case. Seems more likely it would be the result of a buggy application. Interestingly, we have no way to recover from a fatal error here: that is, there is no way for the transaction to roll back. So maybe we have to allow it and log a warning.

hachikuji replied:
Oh actually, that's wrong. We can fail the `TxnOffsetCommit`. I was thinking this was part of the roll forward logic. So then I guess the question is whether it's worthwhile keeping some additional state to be able to tell whether the group has done consumer offset commits for the current generation. If it has, perhaps we should send back an error.

apurvam replied:
I added a check to log a warning if a group has received both types of offset commits in the current instance.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -146,6 +146,7 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   private val members = new mutable.HashMap[String, MemberMetadata]
   private val offsets = new mutable.HashMap[TopicPartition, OffsetAndMetadata]
   private val pendingOffsetCommits = new mutable.HashMap[TopicPartition, OffsetAndMetadata]
+  private val pendingTransactionalOffsetCommits = new mutable.HashMap[Long, mutable.Map[TopicPartition, OffsetAndMetadata]]()
```
Can you document the key type?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -289,6 +290,35 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   }
 
+  def prepareTxnOffsetCommit(producerId: Long, offsets: mutable.Map[TopicPartition, OffsetAndMetadata]) {
+    val producerOffsets = pendingTransactionalOffsetCommits.getOrElseUpdate(producerId, mutable.Map.empty[TopicPartition, OffsetAndMetadata])
```
It might be useful to document the assumptions we are making about the ordering of incoming offset commits. It seems we assume that the client will never have two differing sets of offsets to commit for the same groupId. While this would be super weird, there is nothing that prevents it (that is, unless we add logic to do so). Maybe in that case we just say that the last received commit always wins and if you care about order, you need to block before sending another.

apurvam replied:
This is the existing behavior for non transactional offset commits as well, right?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -289,6 +290,35 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+      pendingTransactionalOffsetCommits.remove(producerId)
+  }
+
+  /* Complete a pending transactional offset commit. This is called when a commit or abort maker is received */
```
typo: "maker"

Also, maybe we can clarify that this is only called after the marker is successfully written to the log?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -289,6 +290,35 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+  }
+
+  /* Complete a pending transactional offset commit. This is called when a commit or abort maker is received */
+  def completePendingTxnOffsetCommit(producerId: Long, isCommit: Boolean): Unit = {
```
Probably would be a good idea to add some trace logging in here.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -289,6 +290,35 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+  def activeProducers = pendingTransactionalOffsetCommits.keySet
+
+  def hasPendingOffsetCommitsFromProducer(producerId: Long) =
+    pendingTransactionalOffsetCommits.getOrElse(producerId, Map.empty[TopicPartition, OffsetAndMetadata]).nonEmpty
```
Isn't this just `pendingTransactionalOffsetCommits.contains(producerId)`? Would we ever have an empty map in `pendingTransactionalOffsetCommits`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -75,6 +75,9 @@ class GroupMetadataManager(brokerId: Int,
   private val scheduler = new KafkaScheduler(threads = 1, threadNamePrefix = "group-metadata-manager-")
 
+  /* The groups with open transactional offsets commits per producer */
+  private val openGroupsForProducer = mutable.HashMap[Long, mutable.Set[String]]()
```
Might be worth a comment mentioning that this is needed to update the group metadata after receiving the transaction marker callback (which has no knowledge of the group). Kind of annoying that we need this by the way. It's more mutable state which needs to be consistent with other mutable state. I wonder if it would be prohibitive to just scan through the active groups when completing the transaction. We could be a little smarter and store the corresponding offsets partition in `GroupMetadata` so that we only touch the groups which could be related.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -263,13 +269,22 @@ class GroupMetadataManager(brokerId: Int,
-          val builder = MemoryRecords.builder(buffer, magicValue, compressionType, timestampType, 0L)
+
+          if (isTxnOffsetCommit && magicValue < RecordBatch.MAGIC_VALUE_V2)
+            throw new IllegalStateException("Attempting to make a transaction offset commit with an invalid magic: " + magicValue)
```
Seems this would result in an UNKNOWN error propagated back to the producer. Maybe instead we could use the UNSUPPORTED_FOR_MESSAGE_FORMAT error code?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -289,6 +290,35 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     pendingOffsetCommits ++= offsets
   }
 
+  def prepareTxnOffsetCommit(producerId: Long, offsets: mutable.Map[TopicPartition, OffsetAndMetadata]) {
```
Does `offsets` need to be mutable?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -441,41 +469,68 @@ class GroupMetadataManager(brokerId: Int,
+                      val value = GroupMetadataManager.readOffsetMessageValue(record.value)
+                      if (isTxnOffsetCommit) {
+                        pendingOffsets(batch.producerId).put(key, value)
+                        addProducerGroup(batch.producerId, key.group)
```
We try to be careful about not modifying any global state until after the partition has been fully read. This ensures that only the most recent state is ever visible. I'm not sure it causes any problems here, but it might be better to stick to the same pattern.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -651,6 +727,36 @@ class GroupMetadataManager(brokerId: Int,
+  }
+
+  private def removeProducerGroup(producerId: Long, groupId: String) = openGroupsForProducer synchronized {
+    openGroupsForProducer.getOrElseUpdate(producerId, mutable.Set.empty[String]).remove(groupId)
```
When do we remove the empty sets from `openGroupsForProducer`?

apurvam replied:
I think I have misunderstood what you are asking. There is a check right below which removes the map entry for the producerId if the set of open groups is empty.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -694,6 +800,8 @@ class GroupMetadataManager(brokerId: Int,
       ownedPartitions.add(partition)
     }
   }
+
```
Was this intentional?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1458,9 +1458,78 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   def handleTxnOffsetCommitRequest(request: RequestChannel.Request): Unit = {
```
Seems like a good portion of the logic below is the same as the normal offset commit handler. Is there any way we can consolidate some of the common logic?

apurvam replied:
You are correct. There are subtle differences in minor things like log messages. I will look into the best way to consolidate.

apurvam replied:
I spent about an hour looking into this. To factor out the common bits in a useful way, I think we will need to make `TxnOffsetCommitRequest` a subclass of `OffsetCommitRequest` (and also the corresponding response classes). 

Otherwise, the same information is represented differently in both, and trying to bridge the difference would be more code than is actually duplicated. I think we should leave the refactor as a future work item.

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments. Looks like you have a bit of a painful rebase first though.

Commit:
Fix build failures and get tests to pass after rebase

Commit:
Addressed review comments:

  1. Main comment was about moving the logic to skip sequence number
checks to the `Log` class. We skip sequence number checks only for the
messages in the __consumer-offsets topic.
  2. Addressed other minor comments.
  3. Added first unit test for loading transactional offset commits.
Still need to add more cases.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -382,8 +382,24 @@ public static MemoryRecordsBuilder builder(ByteBuffer buffer,
+                logAppendTime, producerId, producerEpoch, baseSequence, isTransactional, false, partitionLeaderEpoch);
+    }
+
+    public static MemoryRecordsBuilder builder(ByteBuffer buffer,
```
Do we need this method if all it does is pass the arguments directly to the `MemoryRecordsBuilder` constructor?

apurvam replied:
This method passes in `buffer.remaining()` as an additional parameter to the `MemoryRecordsBuilder` constructor.

hachikuji replied:
Fair enough. We really need to turn these methods into a builder or something.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -287,9 +291,50 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+  def hasPendingOffsetCommitsFromProducer(producerId: Long) =
+    pendingTransactionalOffsetCommits.getOrElse(producerId, Map.empty[TopicPartition, OffsetAndMetadata]).nonEmpty
+
   def removeOffsets(topicPartitions: Seq[TopicPartition]): immutable.Map[TopicPartition, OffsetAndMetadata] = {
```
Maybe we should also remove from `pendingTransactionalOffsetCommits`? I think this is invoked when a partition is deleted.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -287,9 +291,50 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        pendingOffsets.remove(topicPartition)
+        if (pendingOffsets.isEmpty)
+          pendingTransactionalOffsetCommits.remove(producerId)
+      case _ =>
```
Might be worth adding a comment that one situation we might hit this is if the partition is removed by `removeOffsets` below. As far as I can tell, that's the only one.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -264,13 +277,25 @@ class GroupMetadataManager(brokerId: Int,
           // set the callback function to insert offsets into cache after log append completed
           def putCacheCallback(responseStatus: Map[TopicPartition, PartitionResponse]) {
-            // the append response should only contain the topics partition
+           // the append response should only contain the topics partition
```
nit: this was probably unintentional

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -264,13 +277,25 @@ class GroupMetadataManager(brokerId: Int,
+          if (isTxnOffsetCommit && magicValue < RecordBatch.MAGIC_VALUE_V2)
+            throw Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT.exception("Attempting to make a transaction offset commit with an invalid magic: " + magicValue)
+
+          val builder = if (isTxnOffsetCommit)
```
nit: Is this check necessary? Seems like we could follow the first branch in both cases.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -536,6 +623,9 @@ class GroupMetadataManager(brokerId: Int,
     }
     trace(s"Initialized offsets $loadedOffsets for group ${group.groupId}")
     group.initializeOffsets(loadedOffsets)
+    pendingTransactionalOffsets.foreach { case (producerId, offsets) =>
```
nit: maybe we could add `pendingTransactionalOffsets` as a second argument to `GroupMetadata.initialize`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -281,17 +306,23 @@ class GroupMetadataManager(brokerId: Int,
-                  filteredOffsetMetadata.foreach { case (topicPartition, offsetAndMetadata) =>
-                    group.failPendingOffsetWrite(topicPartition, offsetAndMetadata)
-                  }
+                    filteredOffsetMetadata.foreach { case (topicPartition, offsetAndMetadata) =>
```
nit: this block is misaligned

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -281,17 +306,23 @@ class GroupMetadataManager(brokerId: Int,
+                    filteredOffsetMetadata.foreach { case (topicPartition, offsetAndMetadata) =>
+                      if (isTxnOffsetCommit) {
+                        group.failPendingTxnOffsetCommit(producerId, topicPartition, offsetAndMetadata)
+                        if (!group.hasPendingOffsetCommitsFromProducer(producerId))
```
Could we move this check outside the loop?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -434,13 +472,14 @@ class GroupMetadataManager(brokerId: Int,
         while (currOffset < highWaterMark && !shuttingDown.get()) {
           val fetchDataInfo = log.read(currOffset, config.loadBufferSize, maxOffset = None, minOneMessage = true,
-            isolationLevel = IsolationLevel.READ_UNCOMMITTED)
+            isolationLevel = IsolationLevel.READ_COMMITTED)
```
Hmm... why did we change this? I think we're expecting to read even the incomplete transactions so that we can populate the pending transaction offset commits in `GroupMetadata`.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -329,8 +360,15 @@ class GroupMetadataManager(brokerId: Int,
-          group synchronized {
-            group.prepareOffsetCommit(offsetMetadata)
+          if (isTxnOffsetCommit) {
+            addProducerGroup(producerId, group.groupId)
```
Mentioned offline, but adding this to make sure we don't forget about it. I think we need to make sure that we hold the group lock here so that we can keep the `pendingTransactionalOffsetCommits` collection in `GroupMetadata` in sync with `openGroupsForProducer`.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -451,59 +490,106 @@ class GroupMetadataManager(brokerId: Int,
+                      pendingOffsets.put(batch.producerId, mutable.Map[GroupTopicPartition, OffsetAndMetadata]())
+
+                    // load offset
+                    val key = offsetKey.key
```
nit: can we rename this to something more descriptive. Maybe `groupTopicPartition`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -451,59 +490,106 @@ class GroupMetadataManager(brokerId: Int,
+                    val key = offsetKey.key
+                    if (!record.hasValue) {
+                      if (isTxnOffsetCommit)
+                        pendingOffsets(batch.producerId).remove(key)
```
Seems the map could be empty after this remove. In that case, should we remove from `producerGroups`? Either that, or maybe we can check for an empty map before calling `addProducerGroup` below.

apurvam replied:
Actually, I got rid of `producerGroups` all together. I now derive the producer groups from `pendingOffsets` after we load all the data. We are iterating over the `pendingOffsets` map anyway, so I just extract the active groups at that point and call `addProducerGroup` on each.

Commit:
Two updates:

 1. Added serveral test cases for loading the transactions topic with
offset commits.

 2. Added code to invoke the coordinator.handleEndTxn when the
transaction markers have been succesfully written to the log.

Commit:
Address PR Comments

Commit:
Address few more PR comments

apurvam commented:
@hachikuji I addressed most of your comments. I also added test cases for loading offsets topic partitions with transactional offset commits.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -238,12 +250,20 @@ class GroupMetadataManager(brokerId: Int,
       validateOffsetMetadataLength(offsetAndMetadata.metadata)
     }
 
+    if (!group.hasReceivedConsistentOffsetCommits)
```
I think we invoke `prepareStoreOffsets` while holding the group lock, but it might be a good idea to synchronize on the group around this check anyway to make it clear that the group lock is needed.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -238,12 +250,20 @@ class GroupMetadataManager(brokerId: Int,
     }
 
+    if (!group.hasReceivedConsistentOffsetCommits)
+      warn(s"group: ${group.groupId} with leader: ${group.leaderId} has received offset commits from consumers as well" +
```
nit: looks like we're missing a space at the end of "well."

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -565,6 +644,7 @@ class GroupMetadataManager(brokerId: Int,
 
         for (group <- groupMetadataCache.values) {
           if (partitionFor(group.groupId) == offsetsPartition) {
+            group.activeProducers.foreach(removeProducerGroup(_, group.groupId))
```
I don't think accessing `activeProducers` is safe without the group lock. The call to `onGroupUnloaded` will transition the state of the group to `Dead`. Maybe after that, we can just iterate through `openGroupsForProducer` and remove the group from every producer.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -660,6 +740,36 @@ class GroupMetadataManager(brokerId: Int,
+    pendingGroups.foreach { case (groupId) =>
+      getGroup(groupId) match {
+        case Some(group) => group synchronized {
+          group.completePendingTxnOffsetCommit(producerId, isCommit)
```
I think we probably want to check if the group is not dead here.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1434,8 +1434,15 @@ class KafkaApis(val requestChannel: RequestChannel,
+        partitionResponse.error == Errors.NONE
+      }.keys.toSeq
+
+      groupCoordinator.handleTxnCompletion(producerId = pid, topicPartitions = successfulPartitions, transactionResult = result)
```
Discussed offline. At the moment, the call to `handleTxnCompletion` seems safe, but to be resilient to future changes, we should probably add code to catch any exceptions and overwrite the response error codes with an UNKNOWN error. This should cause the TC to retry.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1511,9 +1518,78 @@ class KafkaApis(val requestChannel: RequestChannel,
+    val header = request.header
+    val txnOffsetCommitRequest = request.body[TxnOffsetCommitRequest]
+    // reject the request if not authorized to the group
+    if (!authorize(request.session, Read, new Resource(Group, txnOffsetCommitRequest.consumerGroupId))) {
```
Discussed offline. It might be tough to consolidate all of the logic with the normal offset commit handler, but perhaps we can extract some common functions such as the authorization and topic existence checks.

Commit:
Address remaining PR comments

Commit:
Fail only those partitions for which the offset cache update failed

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

Commit:
Fix compile error with JDK7

Commit:
Drop the errors.contain(pid) check when updating the errors. The pid should always exist, and if not, we should throw an exception

hachikuji commented:
retest this please

guozhangwang commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -281,16 +309,21 @@ class GroupMetadataManager(brokerId: Int,
               } else {
                 if (!group.is(Dead)) {
+                  if (!group.hasPendingOffsetCommitsFromProducer(producerId))
+                    removeProducerGroup(producerId, group.groupId)
```
Hmm, is this a right behavior to clear the producer group if append fails? For example if we have some previous appends for the same group that has been successful and acked, and this append failed due to an error, and the client just gives up and proceed to commit transaction, and then when the commit marker is received we should still reflect the previously acked offset in the cache right? @apurvam @hachikuji

hachikuji replied:
@guozhangwang Hmm.. I'm not sure I understand the case you are talking about. If we had received some previous append from the same producer as part of a transaction, then it would be reflected in the `hasPendingOffsetCommitsFromProducer` call. Can you clarify the case?
